### PR TITLE
THRIFT-4830: Add to_string helper function for cpp generator

### PR DESCRIPTION
**Pull request by soroshsabz**
ITNOA

Hi,

I add to_string helper function for enum in cpp, because it is increase usability and performance.
  

- [x] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  (not required for trivial changes)
- [x] Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?  (not required for trivial changes)
- [x] Did you squash your changes to a single commit?
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"

### CODE REVIEW

**Comment by soroshsabz** [comment]
@jeking3 Please retest this patch, I think CI does not work correctly 

thanks

**Comment by jeking3** [comment]
We have a to_string implementation (not generated by the compiler).  Is it insufficient?  This also needs a unit test in the runtime library for C++.

**Comment by soroshsabz** [comment]
@jeking3 thrift to_string is [based on stringstream](https://github.com/apache/thrift/blob/af7ecd6a2b15efe5c6b742cf4a9ccb31bcc1f362/lib/cpp/src/thrift/TToString.h#L35) but this is poor performance, because creating ostringstream per call is very expensive, so I think generated to_string for enum is good choice.

**Comment by soroshsabz** [comment]
@jeking3 I added unit test for this code :)

**Comment by soroshsabz** [comment]
@jeking3 build failed because Haskell network does not found.

**Comment by soroshsabz** [comment]
@jeking3 Is there any problem yet?

thanks :)

**Comment by soroshsabz** [comment]
@jeking3 I rebase it to latest master

File: compiler/cpp/src/thrift/generate/t_cpp_generator.cc [hunk 1]
```diff
@@ -631,6 +636,47 @@ void t_cpp_generator::generate_enum_ostream_operator(std::ostream& out, t_enum*
...
+    out << "std::string to_string(const ";
+    if (gen_pure_enums_) {
+      out << tenum->get_name();
+    } else {
+      out << tenum->get_name() << "::type&";
+    }
+    out << " val) " ;
+	scope_up(out);
+
+    out << indent() << "std::string out;" << endl;
```
Comment (jeking3) [inline compiler/cpp/src/thrift/generate/t_cpp_generator.cc]: This temporary is unnecessary.

File: compiler/cpp/src/thrift/generate/t_cpp_generator.cc [hunk 2]
```diff
@@ -631,6 +636,47 @@ void t_cpp_generator::generate_enum_ostream_operator(std::ostream& out, t_enum*
...
+    out << indent() << "std::map<int, const char*>::const_iterator it = _"
+             << tenum->get_name() << "_VALUES_TO_NAMES.find(val);" << endl;
+    out << indent() << "if (it != _" << tenum->get_name() << "_VALUES_TO_NAMES.end()) {" << endl;
+    indent_up();
+    out << indent() << "out = std::string(it->second);" << endl;
```
Comment (jeking3) [inline compiler/cpp/src/thrift/generate/t_cpp_generator.cc]: You can return here.

File: compiler/cpp/src/thrift/generate/t_cpp_generator.cc [hunk 3]
```diff
@@ -631,6 +636,47 @@ void t_cpp_generator::generate_enum_ostream_operator(std::ostream& out, t_enum*
...
+    indent_down();
+    out << indent() << "} else {" << endl;
+    indent_up();
+    out << indent() << "out = std::to_string(static_cast<int>(val));" << endl;
```
Comment (jeking3) [inline compiler/cpp/src/thrift/generate/t_cpp_generator.cc]: You can return here.

File: compiler/cpp/src/thrift/generate/t_cpp_generator.cc [hunk 4]
```diff
@@ -631,6 +636,47 @@ void t_cpp_generator::generate_enum_ostream_operator(std::ostream& out, t_enum*
...
+    indent_down();
+    out << indent() << "}" << endl;
+
+    out << indent() << "return out;" << endl;
```
Comment (jeking3) [inline compiler/cpp/src/thrift/generate/t_cpp_generator.cc]: out is unnecessary

**Comment by Jens-G** [comment]
@soroshsabz: any news here?

**Comment by soroshsabz** [comment]
@Jens-G sorry I had a work problem with me in past month, and so I am very busy.

I resolved all @jeking3 review changes request.

thanks to @jeking3 for review my PR :)

**Pull request merged**